### DISCUSSION TITLE
Borderless maximized monocle?

### DISCUSSION THREAD

**Discussion by gekoke** [leftwm/leftwm#745] [Q&A]
Hi, awesome project - really fun to configure and everything works easily out of the box!

Coming from a patched dwm, one of the only things I miss is a monocle mode where the borders are removed, yet the bar remains visible (I've attached a picture to illustrate). This was already discussed in #174, but I believe only a fullscreen mode was implemented since then.

This would be different from the fullscreen mode in that:
- any bars would remain visible
- windows would remain cycleable

Is there any way to have this functionality that I've missed? Essentially, I wouldn't mind having the border and gap settings be different between monocle and tiled layouts either - is there some way to do _this_ perhaps that I've missed?

Currently I use a script to toggle the gap size per tag, though I haven't found a way to toggle the borders either (perhaps it's undocumented or I just missed it?).

Thanks!

![2022-03-26_01-25](https://user-images.githubusercontent.com/63151928/160214257-73ed5bb4-512b-40ad-8ee9-30477a29cf7e.png)

**Comment by M-Mabrouk1** [comment]
I think this will be a neat feature that the window will be no gaps if there is only one window in the workspace.
can you tell me how you were able to do it with a script?

**Comment by VuiMuich** [comment]
Just had an idea for a workaround, but as most workarounds it is not perfect so a feature to toggle borders might indeed be a nice feature.
I believe we have a wish for that already tracked somewhere. Which reminds me, some issue chores are a bit overdue, I guess...

Now the workaround is, to define a second `theme.toml` and either use a script that reads the current layout from `leftwm state` and loads either the `monocle-theme.toml` or restores the regular one.

**Comment by gekoke** [comment]
@M-Mabrouk1 Here's the script I used to toggle the gap size per workspace. I don't think there's a way to do something similar for the borders, unfortunately, so you would have to swap the theme as VuiMuich mentioned.

```py
#!/usr/bin/env python3

import shelve
import json
import subprocess
import os
import appdirs


CACHE_DIR = appdirs.user_cache_dir("toggle-fullscreen-leftwm-monocle")
CACHE_FILE = os.path.join(CACHE_DIR, "cache.db")


def ensure_cache() -> None:
    os.makedirs(os.path.expanduser(CACHE_DIR), exist_ok=True)


def get_active_tag() -> int:
    state = json.loads(subprocess.run(["leftwm", "state", "-q"], capture_output=True).stdout)
    tags = state["workspaces"][0]["tags"]
    for i, tag in enumerate(tags):
        if tag["focused"]:
            return i
    raise Exception("No active tag found")


def toggle_fullscreen_monocle(tag_index: int) -> None:
    with shelve.open(CACHE_FILE) as cache:
        key = f"tag_{tag_index}_maximized_last"
        if key not in cache:
            cache[key] = False
        elif cache[key]:
            subprocess.run(["leftwm", "command", "SetMarginMultiplier 1"])
            cache[key] = False
        else:
            subprocess.run(["leftwm", "command", "SetMarginMultiplier 0"])
            cache[key] = True
    

def main() -> None:
    ensure_cache()
    toggle_fullscreen_monocle(get_active_tag())


if __name__ == "__main__":
    main()
```
